### PR TITLE
Add JSON API basic support (continuation)

### PR DESCRIPTION
**Pull request by meyerbaptiste**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | api-platform/api-platform#82
| License       | MIT
| Doc PR        | N/A

Continuation (and conclusion?) of PRs #785 and #1036.

First, I rebase the branch against the master branch (:cold_sweat:).
Then (see the **second** commit):
* I fixed some issues according to the JSON API schema definition,
* I fixed Behat tests,
* I added some PHPUnit tests,
* I added the enabling of the PATCH operation only if JSON API support is enabled,
* I removed dead code.

GTK:
* This implementation is very, very basic,
* IMO, normalizers could be refactored in the future.

Did I forget something?

Cc: @hectorh30, @Simperfit

### CODE REVIEW

File: src/JsonApi/EventListener/TransformSortingParametersListener.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+        // This applies only to jsonapi request format
+        if ('jsonapi' !== $request->getRequestFormat()) {
+            return;
+        }
+
+        // If order query parameter is not defined or is already an array, never mind
+        $orderParameter = $request->query->get($this->orderParameterName);
+        if (null === $orderParameter || is_array($orderParameter)
+        ) {
```
Comment (Simperfit) [inline src/JsonApi/EventListener/TransformSortingParametersListener.php]: strange cs?

**Comment by soyuka** [comment]
> First, I rebase the branch against the master branch (:cold_sweat:).

Wow! For this you deserve much :heart:! To bad that authors history was lost though. Really nice if this can finally come to an end :D. Thanks @meyerbaptiste!

**Comment by Simperfit** [comment]
@soyuka Yep, anyway the author are in the files and we could `@them` when mergin :).

Thanks @meyerbaptiste for finishing this !

File: src/JsonApi/EventListener/TransformSortingParametersListener.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Converts pagination parameters from JSON API recommended convention to
+ * api-platform convention.
+ *
+ * @see http://jsonapi.org/format/#fetching-sorting
+ * @see https://api-platform.com/docs/core/filters#order-filter
+ *
+ * @author Héctor Hurtarte <hectorh30@gmail.com>
+ */
+final class TransformSortingParametersListener
```
Comment (soyuka) [inline src/JsonApi/EventListener/TransformSortingParametersListener.php]: It feels like those `parametersListener` have a lot of code duplication. I don't know if it's okay like this or if we should base those on an abstract class. WDYT?

File: src/JsonApi/Serializer/CollectionNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    public function normalize($data, $format = null, array $context = [])
+    {
+        $currentPage = $lastPage = $itemsPerPage = 1;
+
+        // If we are normalizing stuff one level down (i.e., an attribute which
+        // could be already an array)
+        $returnDataArray = [];
+        if (isset($context['api_sub_level'])) {
+            foreach ($data as $index => $obj) {
+                $returnDataArray['data'][][$index] = $this->normalizer->normalize($obj, $format, $context);
```
Comment (soyuka) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: Shouldn't the `$returnDataArray` be like `$returnDataArray = ['data' => []]` on initialization for this to work? I don't get how `['data' => [[]]]` is self-initializing.

File: src/JsonApi/Serializer/CollectionNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+        $parsed = IriHelper::parseIri($context['uri'] ?? '/', $this->pageParameterName);
+        $paginated = $isPaginator = $data instanceof PaginatorInterface;
+
+        if ($isPaginator) {
+            $currentPage = $data->getCurrentPage();
+            $lastPage = $data->getLastPage();
+            $itemsPerPage = $data->getItemsPerPage();
+
+            $paginated = 1. !== $lastPage;
```
Comment (soyuka) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: `1.` ? Because of float comparison? Saw that below too.

File: src/JsonApi/Serializer/ItemNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+                    $isMany = null !== $valueType && ($className = $valueType->getClassName()) && $this->resourceClassResolver->isResourceClass($className);
+                } else {
+                    $className = $type->getClassName();
+
+                    $isOne = null !== $className && $this->resourceClassResolver->isResourceClass($className);
+                }
+
+                $typeShortName = '';
+
+                if ($className && $this->resourceClassResolver->isResourceClass($className)) {
```
Comment (soyuka) [inline src/JsonApi/Serializer/ItemNormalizer.php]: Because `isResourceClass` is quite heavy (#1068) I'd prefer if we could use a variable to store it above instead of calling it a least twice.

File: src/JsonApi/Serializer/ItemNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+
+    /**
+     * Gets the cache key to use.
+     *
+     * @param string|null $format
+     * @param array       $context
+     *
+     * @return bool|string
+     */
+    private function getJsonApiCacheKey(string $format = null, array $context)
```
Comment (soyuka) [inline src/JsonApi/Serializer/ItemNormalizer.php]: method typing `:bool`?

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 7]
```diff
@@ -183,6 +183,8 @@ private function getPathOperation(string $operationName, array $operation, strin
                 return $this->updateGetOperation($pathOperation, $mimeTypes, $operationType, $resourceMetadata, $resourceClass, $resourceShortName, $operationName, $definitions);
             case 'POST':
                 return $this->updatePostOperation($pathOperation, $mimeTypes, $operationType, $resourceMetadata, $resourceClass, $resourceShortName, $operationName, $definitions);
+            case 'PATCH':
+                $pathOperation['summary'] ?? $pathOperation['summary'] = sprintf('Updates the %s resource.', $resourceShortName);
```
Comment (soyuka) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: interesting hack :stuck_out_tongue: we could `break` after this IMO so that it gets to the return statement directly.

File: tests/Hydra/Serializer/ItemNormalizerTest.php [hunk 8]
```diff
@@ -44,12 +44,10 @@ public function testDontSupportDenormalization()
         $contextBuilderProphecy = $this->prophesize(ContextBuilderInterface::class);
         $resourceClassResolverProphecy->getResourceClass(['dummy'], 'Dummy')->willReturn(Dummy::class);
         $propertyNameCollectionFactoryProphecy->create(Dummy::class, [])->willReturn(new PropertyNameCollection(['name' => 'name']));
-        $propertyMetadataFactoryProphecy->create(Dummy::class, 'name', [])->willReturn(new PropertyMetadata())->shouldBeCalled(1);
 
         $normalizer = new ItemNormalizer($resourceMetadataFactoryProphecy->reveal(), $propertyNameCollectionFactoryProphecy->reveal(), $propertyMetadataFactoryProphecy->reveal(), $iriConverterProphecy->reveal(), $resourceClassResolverProphecy->reveal(), $contextBuilderProphecy->reveal());
 
         $this->assertFalse($normalizer->supportsDenormalization('foo', ItemNormalizer::FORMAT));
-        $normalizer->denormalize(['foo'], Dummy::class, 'jsonld', ['jsonld_has_context' => true, 'jsonld_sub_level' => true, 'resource_class' => Dummy::class]);
```
Comment (soyuka) [inline tests/Hydra/Serializer/ItemNormalizerTest.php]: why did you drop this test?

File: tests/JsonApi/EventListener/FlattenPaginationParametersListenerTest.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        $request = new Request(['page' => ['size' => 5, 'number' => 3]]);
+        $request->setRequestFormat('jsonapi');
+
+        $eventProphecy = $this->prophesize(GetResponseEvent::class);
+        $eventProphecy->getRequest()->willReturn($request)->shouldBeCalled();
+
+        $this->listener->onKernelRequest($eventProphecy->reveal());
+
+        $expectedRequest = new Request(['page.size' => 5, 'page.number' => 3]);
```
Comment (soyuka) [inline tests/JsonApi/EventListener/FlattenPaginationParametersListenerTest.php]: we could also test negative page parameters values to avoid a recent issue we had.

**Comment by Simperfit** [comment]
@meyerbaptiste could you push it a little bit further by adding that 0.6% coverage missing ?

File: src/JsonApi/Serializer/CollectionNormalizer.php (hunk 4 above)
Comment (meyerbaptiste) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: Yes, I think. `$lastPage` is a float.

File: src/JsonApi/Serializer/ItemNormalizer.php (hunk 6 above)
Comment (meyerbaptiste) [inline src/JsonApi/Serializer/ItemNormalizer.php]: ? This method returns a string or false.

File: src/Swagger/Serializer/DocumentationNormalizer.php (hunk 7 above)
Comment (meyerbaptiste) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Why adding a `break` here? I have to return in the `PUT` case.

Comment (soyuka) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Oh okay then thought it returned as default.

File: src/JsonApi/Serializer/ItemNormalizer.php (hunk 6 above)
Comment (soyuka) [inline src/JsonApi/Serializer/ItemNormalizer.php]: wow, my bad missed that.

File: src/JsonApi/Serializer/CollectionNormalizer.php (hunk 4 above)
Comment (soyuka) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: https://github.com/api-platform/core/pull/1175/files#diff-5b154fa3efaf5776695baccd0a52daf4R66 initialized as integer? IMO pagination stuff should be integer-only because it doesn't make sense to have a float ^^ (maybe jsonapi specifies float though I haven't looked at it).

File: tests/Hydra/Serializer/ItemNormalizerTest.php (hunk 8 above)
Comment (meyerbaptiste) [inline tests/Hydra/Serializer/ItemNormalizerTest.php]: Huum, I think it was because the test is about denormalization support and not denormalization.

File: tests/JsonApi/EventListener/FlattenPaginationParametersListenerTest.php (hunk 9 above)
Comment (meyerbaptiste) [inline tests/JsonApi/EventListener/FlattenPaginationParametersListenerTest.php]: Okay, but at this level, positive or negative, it has no impact.

File: tests/Hydra/Serializer/ItemNormalizerTest.php (hunk 8 above)
Comment (soyuka) [inline tests/Hydra/Serializer/ItemNormalizerTest.php]: Yeah it's weird indeed, test is `testDontSupportDenormalization`. Then we still do the denormalization with no assertions ?!

Still, `denormalize` isn't tested anymore in this class, maybe we could move it in a new test with an assertion then?

File: src/Swagger/Serializer/DocumentationNormalizer.php (hunk 7 above)
Comment (meyerbaptiste) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: https://github.com/api-platform/core/pull/1036/files#diff-32d63273c86a7e4c14c3fa6bb4131194

Comment (soyuka) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: :+1: better like this indeed.

File: src/JsonApi/Serializer/CollectionNormalizer.php (hunk 4 above)
Comment (meyerbaptiste) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: I think they wanted to keep a consistency with the `PaginatorInterface` interface which returns floats.

File: src/JsonApi/Serializer/CollectionNormalizer.php (hunk 3 above)
Comment (meyerbaptiste) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: The magic of PHP: https://3v4l.org/GXaOT

Comment (soyuka) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: Wow okay, I feel dumb right now, thanks @meyerbaptiste!

File: tests/Hydra/Serializer/ItemNormalizerTest.php (hunk 8 above)
Comment (meyerbaptiste) [inline tests/Hydra/Serializer/ItemNormalizerTest.php]: Yep, you're right.

> Then we still do the denormalization with no assertions

It's wrong, because `shouldBeCalled()` is an assertion 😛

Comment (soyuka) [inline tests/Hydra/Serializer/ItemNormalizerTest.php]: haha yeah I meant no assertion on the `denormalization` call like we do on `normalization` below.

File: src/JsonApi/Serializer/ItemNormalizer.php (hunk 5 above)
Comment (meyerbaptiste) [inline src/JsonApi/Serializer/ItemNormalizer.php]: I refactored this part.

File: src/JsonApi/Serializer/CollectionNormalizer.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+     */
+    public function normalize($data, $format = null, array $context = [])
+    {
+        $currentPage = $lastPage = $itemsPerPage = 1.;
+
+        // If we are normalizing stuff one level down (i.e., an attribute which could be already an array)
+        $returnDataArray = [];
+        if (isset($context['api_sub_level'])) {
+            foreach ($data as $index => $obj) {
+                $returnDataArray['data'][][$index] = $this->normalizer->normalize($obj, $format, $context);
```
Comment (meyerbaptiste) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: I'm really not confident with this piece of code. I don't think it works properly but at the same time, I don't know what is the correct result. I am not familiar with sub-resources...
Anyone can help me here?

Comment (soyuka) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: Looks correct according to what are the other normalizer's doing in case of a `api_sub_level` flag.
https://github.com/api-platform/core/blob/b88d975fa152d295ffeed332dab81dd8376722d1/src/Hal/Serializer/CollectionNormalizer.php#L60
https://github.com/api-platform/core/blob/183b0a58114a3294380c27c6ff7b2536b128523a/src/Hydra/Serializer/CollectionNormalizer.php#L65

Comment (meyerbaptiste) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: It's more about the array structure. What type of values can be the index and the normalized object here? According to JSON API spec, `data` must be primary data or represent resource linkage, we can not put what we want in it! :confused:

**Comment by meyerbaptiste** [comment]
I got +0.1% @Simperfit! I don't want to write tests for at least 1 month...

**Comment by Simperfit** [comment]
it's good for me ;)

**Comment by hectorh30** [comment]
Awesome! Thanks so much for going on with this. I also went out of time to complete this, but I'm glad to see it moving forward.

File: src/JsonApi/Serializer/CollectionNormalizer.php (hunk 4 above)
Comment (dunglas) [inline src/JsonApi/Serializer/CollectionNormalizer.php]: Indeed. We use float because of large numbers.

**Comment by meyerbaptiste** [comment]
I'll add an automatic reminder to rebase every week! 😄

**Comment by StephenOTT** [comment]
Is there major work on this still to be done?  Blockers?

**Comment by Simperfit** [comment]
@StephenOTT This will not be part of the 2.1 release. It will be released after.

**Comment by StephenOTT** [comment]
Sure.  But is there still work to be done.  We have devs looking for these features and are ready to contribute.  Just need some direction on whats left to do.

**Comment by hectorh30** [comment]
@Simperfit do you know of any plan to include this in a following release?

**Comment by meyerbaptiste** [comment]
Hi @hectorh30, I still have some tests to write/fix, I made some corrections recently.

**Comment by meyerbaptiste** [comment]
I think it's ready. Can I have a last revue @api-platform/core-team?

File: features/bootstrap/JsonApiContext.php [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+    /**
+     * Checks that given JSON node is equal to an empty array.
+     *
+     * @Then the JSON node :node should be an empty array
+     */
+    public function theJsonNodeShouldBeAnEmptyArray($node)
+    {
+        if (!is_array($actual = $this->getValueOfNode($node)) || !empty($actual)) {
+            throw new \Exception(
```
Comment (dunglas) [inline features/bootstrap/JsonApiContext.php]: Can be inlined

File: features/bootstrap/JsonApiContext.php [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                sprintf("The node value is '%s'", json_encode($actual))
```
Comment (dunglas) [inline features/bootstrap/JsonApiContext.php]: Can you single quotes (weird that PHP CS Fixer doesn't detect this)

File: src/JsonApi/EventListener/TransformFieldsetsParametersListener.php [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * @see http://jsonapi.org/format/#fetching-sparse-fieldsets
+ * @see https://api-platform.com/docs/core/filters#property-filter
+ *
+ * @author Baptiste Meyer <baptiste.meyer@gmail.com>
+ */
+class TransformFieldsetsParametersListener
```
Comment (dunglas) [inline src/JsonApi/EventListener/TransformFieldsetsParametersListener.php]: Can't be `final`?

File: src/Serializer/AbstractItemNormalizer.php [hunk 14]
```diff
@@ -435,7 +427,7 @@ protected function getAttributeValue($object, $attribute, $format = null, array
     }
 
     /**
-     * Normalizes a relation as an URI if is a Link or as a JSON-LD object.
+     * Normalizes a relation as a JSON-LD object if is a Link or as an URI.
```
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php]: I think we can remove JSON-LD now :)

**Comment by meyerbaptiste** [comment]
Thanks @Simperfit & @hectorh30, 4 Oct 2016 -> 13 Oct 2017, 1 year of collaborative work!

**Comment by Simperfit** [comment]
Thank you for finishing it !

**Comment by soyuka** [comment]
It's MERGGEEEEDDDD :heart: :heart: :heart: 
big thanks @Simperfit @meyerbaptiste @hectorh30 and every one who reviewed this!

**Comment by StephenOTT** [comment]
@mario-digitalstate

**Comment by jonday-simitive** [comment]
Well done to everyone that got helped out with this. Looks really great, been testing it out since it was merged in and great to see it has made it into a release. Is there any plans on getting include working http://jsonapi.org/format/#fetching-includes ... If not I'd be interested in contributing as it is one of my favorite parts of JSON API.

**Comment by soyuka** [comment]
This looks really similar to what our Property and Group filters are doing. See https://github.com/api-platform/docs/blob/master/core/filters.md#serializer-filters

**Comment by jonday-simitive** [comment]
@soyuka I could be wrong but it seems like the current implementation of JSON API will only provide me with the relationships property in the response if I add a related entity with a Group annotation for GET for example

```
{
  "links": {
    "self": "/api/cars/2"
  },
  "meta": {
    "totalItems": 1,
    "itemsPerPage": 25,
    "currentPage": 1
  },
  "data": [
    {
      "id": "/api/cars/2",
      "type": "Car",
      "attributes": {
        "model": "Fiesta",
        "_id": 2,
      },
      "relationships": {
        "modal": {
          "data": {
            "type": "Modal",
            "id": "/api/modal/77"
          }
        }
      }
    }
  ]
}
```

So a second API request need to be made. With the functionality that has yet to be implemented of the "include" query string to allow you to return:

```
{
  "links": {
    "self": "/api/cars/2?include=make"
  },
  "meta": {
    "totalItems": 1,
    "itemsPerPage": 25,
    "currentPage": 1
  },
  "data": [
    {
      "id": "/api/cars/2",
      "type": "Car",
      "attributes": {
        "model": "Fiesta",
        "_id": 2,
      },
      "relationships": {
        "modal": {
          "data": {
            "type": "Modal",
            "id": "/api/modal/77"
          }
        }
      }
    }
  ],
  "included": [
    {
      "type": "modal",
      "id": "88",
      "attributes": {
        "name": "Ford",
      }
    }
  ]
}
```

**Comment by soyuka** [comment]
Indeed, it's just very similar and the implementation of `include` could follow the same logic as in those filters :).

**Comment by jonday-simitive** [comment]
@soyuka Ah yeah I get you, sorry been a long day ... thought you were saying I could use them already to get the included functionality. I'll take a look at those now though, thanks.

**Comment by cjunge-work** [comment]
Just a note on the includes. They make caching really difficult. A change in an `include`d entity requires any caching to be purged for all other entities that `include` it. IMHO it's one of the flaws with JSON-API and similar formats that support includes.

A better option is to adopt HTTP/2 for faster connections with less latency and cache at the individual entity level (aggregate in DDD).

`Include`d entities is primarily about saving requests, but HTTP/2 makes requests concurrent and faster/lighter than HTTP/1. `Include`s make the responses bigger and harder to maintain and less cacheable.

Making the JSON-API support feature-complete compared to the spec is a good thing, but I thought I'd bring some attention to the issues associated with using this feature. My opinion is based on using includes in a project, and having some big issues.

**Comment by bangbambang** [comment]
@cjunge-work I think it's a common issue with embedded relations (i.e using serialization group), irrespective of the format used. And IMO the best way to deal with it is to be mindful on what to embed for each responses to prevent overcomplicating cache management logic. 

I personally hope this functionality would be available to provide consistency when used with content negotiation.

**Comment by deivid11** [comment]
Hey thanks for the feature! How to enable this json api support?

**Comment by Ronnie-J** [comment]
@jonday-simitive did you start working on this? I am in a situation where I need a subresource to be included with data.

**Comment by jonday-simitive** [comment]
@Ronnie-J We ended up using multiple API requests or the SubResources Api Platform functionality. I've been trying to find the time to implement it but been unable to so far :(
